text/0002-preload-realm.md

```diff
@@ -0,0 +1,479 @@
+is called when a service worker context will be evaluated. A preload realm will be created in
+response to this hook.
+
+#### ShadowRealm proposal
+
+[ShadowRealms](https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md) are an
+ECMAScript TC39 [Stage 3](https://tc39.es/process-document/) proposal which introduce the ability to
+create new isolated contexts from any existing JS context. Although not yet finalized, there's an
+implementation available in V8 and Chromium which can be used as a reference.
+
+> [!NOTE] It's possible to test ShadowRealms in Electron by adding a V8 flag:
+>
+> ```js
+> app.commandLine.appendSwitch("js-flags", "--harmony-shadow-realm");
+> ```
+
+In Chromium, an [implementation for a V8 hook named `OnCreateShadowRealmV8Context` is
+provided.](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/bindings/core/v8/shadow_realm_context.cc;l=63;drc=f092aac6779095fcb906f43fdcd223cd7148483f)
+In this hook, a new `v8::Context` is created with a `ShadowRealmGlobalScope` blink execution
+context. Using blink APIs would allow the context to interact with debugger agents to provide a
+DevTools debugging experience-although not currently exposed in Chrome.
+
+The implementation of ShadowRealms provides a basis for the design of Preload Realms. The
+`ShadowRealmGlobalScope` can be reused for Preload Realms or copied and renamed with non-significant
+duplication.
+
+### Preload realm scripts
+
+To make preload realms useful for app developers, they'll need to be able to run their own scripts
+in its context.
+
+For Electron's existing preload scripts, this is handled in sandboxed renderers by using the
+[synchronous `BROWSER_SANDBOX_LOAD`
+IPC](https://github.com/electron/electron/blob/3609fc7402881b1d51f6c56249506d5dd3fcbe93/lib/sandboxed_renderer/init.ts#L34)
+to gather scripts.
+
+The same approach can be used by Preload Realms. A new initialization script will be added in
+`lib/preload_realm/` to invoke an IPC to collect scripts intended for service workers.
+
+To register scripts, `Session#setPreloads` can be expanded to support multiple target contexts.
+
+```ts
+session.defaultSession.setPreloads([
+  // Existing usage -- maintained for backwards compat
+  "frame-preload.js",
+  // New equivalent usage
+  {
+    type: "frame",
+    script: "frame-preload.js",
+  },
+  // New context type
+  {
+    type: "service-worker",
+    script: "sw-preload.js",
+  },
+]);
+```
+
+### `contextBridge`
+
+A service worker's preload script should have access to the initiator context from the preload realm
+context. The goal of bridging objects between two contexts remains the same, but may require new
+terminologies. A "main world" might not make sense for workers so I've elected to use "initiator
```

Comment:
I much prefer 'initiator' over 'main' in the current context as well (with the node part often also being called main) so I'd be in favor of this terminology change in general.